### PR TITLE
add fallback constants for Argon2

### LINKED ISSUE #183: FFI::NotFoundError: Function 'crypto_pwhash_memlimit_min' not found in [libsodium.so]
I'm deploying on Debian 8 (Jessie) with latest updates and libsodium installed from jessie-backports, 

The version of libsodium that is installed is 1.0.11 (https://packages.debian.org/source/jessie-backports/libsodium)

I get this error when I try to require rbnacl :
`FFI::NotFoundError: Function 'crypto_pwhash_memlimit_min' not found in [libsodium.so]`

Do you know how to fix that ?

### PR DESCRIPTION
**Pull request by valthon**
fixes #183

### CODE REVIEW

File: lib/rbnacl/sodium.rb [hunk 1]
```diff
@@ -29,13 +29,19 @@ def primitive
...
+    def sodium_constant(constant, name: constant, fallback: nil)
       fn = if sodium_primitive
              "crypto_#{sodium_type}_#{sodium_primitive}_#{constant.to_s.downcase}"
            else
              "crypto_#{sodium_type}_#{constant.to_s.downcase}"
            end
-      attach_function fn, [], :size_t
+      begin
+        attach_function fn, [], :size_t
+      rescue FFI::NotFoundError > nfe
```
[the remark this answers is not in the document]
Comment (valthon) [reply lib/rbnacl/sodium.rb]: Thanks!  It should, indeed.

**Comment by valthon** [comment]
Anything else needed for this to get merged?  The travis-ci error has nothing to do with this pull request.

@tarcieri tagging you since you were active on the related issue

File: lib/rbnacl/sodium.rb [hunk 2]
```diff
@@ -29,13 +29,19 @@ def primitive
...
              "crypto_#{sodium_type}_#{constant.to_s.downcase}"
            end
-      attach_function fn, [], :size_t
+      begin
+        attach_function fn, [], :size_t
+      rescue FFI::NotFoundError => nfe
+        raise nfe if fallback.nil?
+
+        define_singleton_method fn, -> { fallback }
+      end
```
[the remark this answers is not in the document]
Comment (valthon) [reply lib/rbnacl/sodium.rb]: hopefully this new version is cleaner, now.  I split out the `fn_name` construction more explicitly and took advantage of the default `raise`.  i didn't switch to your `if fallback` pattern for two reasons: 1) wanted to be able to specify a constant that is set to `false` and 2) rubocop generally tells me to use the guard pattern for this sort of logic and it reads better to me.

File: lib/rbnacl/sodium.rb [hunk 3]
```diff
@@ -29,14 +29,19 @@ def primitive
...
-      fn = if sodium_primitive
-             "crypto_#{sodium_type}_#{sodium_primitive}_#{constant.to_s.downcase}"
-           else
-             "crypto_#{sodium_type}_#{constant.to_s.downcase}"
-           end
-      attach_function fn, [], :size_t
-      const_set(name, public_send(fn))
+    def sodium_constant(constant, name: constant, fallback: nil)
+      fn_name_components = ["crypto", sodium_type, sodium_primitive, constant.to_s.downcase]
+      fn_name = fn_name_components.reject(&:nil?).join("_")
```
Comment (tarcieri) [inline lib/rbnacl/sodium.rb]: Super minor nit: `reject(&:nill?)` can be replaced with `compact` (surprised RuboCop doesn't lint that?)

**Comment by tarcieri** [comment]
Thank you! I'll try to get a release out soon.

**Pull request merged**